lib/lpc55-flash/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+    #[allow(improper_ctypes)]
+    static __this_image: ();
+}
+
+macro_rules! this_image {
+    () => {
+        core::ptr::addr_of!(__this_image) as u32
+    };
+}
+
+macro_rules! image_a_base {
+    () => {
+        core::ptr::addr_of!(__IMAGE_A_BASE) as u32
+    };
+}
+
+macro_rules! image_a_end {
+    () => {
+        core::ptr::addr_of!(__IMAGE_A_END) as u32
+    };
+}
+
+macro_rules! image_b_base {
+    () => {
+        core::ptr::addr_of!(__IMAGE_B_BASE) as u32
+    };
+}
+
+macro_rules! image_b_end {
+    () => {
+        core::ptr::addr_of!(__IMAGE_B_END) as u32
+    };
+}
+
+macro_rules! image_stage0_base {
+    () => {
+        core::ptr::addr_of!(__IMAGE_STAGE0_BASE) as u32
+    };
+}
+
+macro_rules! image_stage0_end {
+    () => {
+        core::ptr::addr_of!(__IMAGE_STAGE0_END) as u32
+    };
+}
+
+fn get_base(which: ImageType) -> u32 {
+    match which {
+        ImageType::ImageA => unsafe { image_a_base!() },
+        ImageType::ImageB => unsafe { image_b_base!() },
+        ImageType::Bootloader => unsafe { image_stage0_base!() },
+    }
+}
+
+fn get_end(which: ImageType) -> u32 {
+    match which {
+        ImageType::ImageA => unsafe { image_a_end!() },
+        ImageType::ImageB => unsafe { image_b_end!() },
+        ImageType::Bootloader => unsafe { image_stage0_end!() },
+    }
+}
+
+fn same_image(which: ImageType) -> bool {
+    get_base(which) == unsafe { this_image!() }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn __write_block(
+    image_num: ImageType,
+    page_num: u32,
+    buffer: *mut u32,
+) -> HypoStatus {
+    // Can only update opposite image
+    if same_image(image_num) {
+        return HypoStatus::SameImage;
+    }
+
+    let base = get_base(image_num);
+    // This is the safe way to write
+    let write_addr = if let Some(page) =
+        page_num.checked_mul(lpc55_romapi::FLASH_PAGE_SIZE as u32)
+    {
+        if let Some(addr) = page.checked_add(base) {
+            addr
+        } else {
+            return HypoStatus::OutOfBounds;
+        }
+    } else {
+        return HypoStatus::OutOfBounds;
+    };
+
+    // Check that the end is within our image
+    if let Some(end) = write_addr.checked_add(FLASH_PAGE_SIZE as u32) {
+        if end > get_end(image_num) {
+            return HypoStatus::OutOfBounds;
+        }
+    } else {
+        return HypoStatus::OutOfBounds;
+    }
+
+    // We expect this to be called from non-secure (running on 28) and
+    // non-privileged mode (called from hubris task). The tt instructions
+    // are mostly useless for doing any kind of checking on the buffer
+    // address passed in. The failure mode is going to be a fault.
+
+    if let Err(result) = flash_erase(write_addr, FLASH_PAGE_SIZE as u32) {
+        return HypoStatus::FlashError(result);
+    }
+
+    if let Err(result) = flash_write(write_addr, buffer, FLASH_PAGE_SIZE as u32)
+    {
+        return HypoStatus::FlashError(result);
+    }
+
+    return HypoStatus::Success;
```

Comment:
FWIW, this function contains a lot of noise because it isn't returning `Result`, so you can't use `?` and its automatic conversions. Consider factoring the body of the function into a routine using `Result`. You can then do things like chained checked arithmetic as

```rust
let x = y.checked_mul(z).and_then(|product| product.checked_add(base)).ok_or(Whatever)?;
```

You could then call that routine from this `extern "C"` routine and convert the result.